pbmtools/pbmtojbig.1
.TH PBMTOJBIG 1 "1995-05-28"
.SH NAME
pbmtojbig \- portable bitmap to JBIG file converter
.SH SYNOPSIS
.B pbmtojbig
[
.I options
]
[
.I input-file
| \-  [
.I output-file
]]
.br
.SH DESCRIPTION
Reads in a portable bitmap (PBM)
from a file or standard
input, compresses it and outputs the image as a
.I JBIG
bi-level image entity (BIE) file.

.I JBIG
is a highly effective lossless compression algorithm for
bi-level images (one bit per pixel) which is particularly suitable
for scanned document pages.

A
.I JBIG
encoded image can be stored in several resolutions (progressive mode).
These resolution layers can be stored all in one singe BIE or they
can be stored in several separate BIE files.
All resolution layers except the lowest one are stored merely as
differences to the next lower resolution layer, because this requires less
space than encoding the full image completely every time. Each resolution
layer has twice the number of horizontal and vertical pixels than
the next lower layer.
.I JBIG
files can also store several bits per pixel as separate bitmap planes,
however this program stores only one single plane.

.SH OPTIONS
.TP 14
.BI \-
A single hyphen instead of an input file name will cause 
.I jbigtopbm
to read the data from standard input instead from a file.
.TP
.BI \-q
Encode the image in one single resolution layer (sequential mode). This
is usually the most efficient compression method. By default, the number
of resolution layers is chosen automatically so that the lowest layer
image is not larger than 640 \(mu 480 pixels.
.TP
.BI \-x " number"
Specify the maximal horizontal size of the lowest resolution layer.
The default is 640 pixels.
.TP
.BI \-y " number"
Specify the maximal vertical size of the lowest resolution layer.
The default is 480 pixels.
.TP
.BI \-l " number"
Select the lowest resolution layer that will be written to the
BIE. It is possible to store the various resolution layers of a 
.I JBIG
image in progressive mode into different BIEs. Options
.BI \-l
and
.BI \-h
allow to select the resolution layer intervall that will appear
in the created BIE. The lowest resolution layer has number 0 and
this is also the default value. By default all layers will be written.
.TP
.BI \-h " number"
Select the highest resolution layer that will be written to the
BIE. By default all layers will be written. See also option
.BI \-l.
.TP
.BI \-d " number"
Specify the total number of differential resolution layers into which the
input image will be split in addition to the lowest layer. Each additional
layer reduces the size of layer 0 by 50 %. This option overrides options
.BI \-x
and
.BI \-y
which are usually a more comfortable way of selecting the number of
resolution layers.
.TP
.BI \-s " number"
The
.I JBIG
algorithm splits each image into a number of horizontal stripes. This
option specifies that each stripe shall have
.I number
lines in layer 0. The default value is selected so that approximately
35 stripes will be used for the whole image.
.TP
.BI \-m " number"
Select the maximum horizontal offset of the adaptive template pixel.
The
.I JBIG
encoder uses a number of neigbour pixels in order to get statistical
a priori knowledge of the propability, whether the next pixel will be black
or white. One single pixel out of this template of context neighbour
pixels can be moved around. Especially for dithered images it can be a
significant advantage to have one neighbour pixel which has a distance large
enough to cover the period of a dither function. By default, the
adaptive template pixel can be moved up to 8 pixels away. This encoder
supports up to 23 pixels, however as decoders are only required to
support at least a distance of 16 pixels by the standard, no higher value
than 16 for
.I number
is recommended in order to maintain interoperability with other
.I JBIG
implementations. The maximal vertical offset of the adaptive
template pixel is always zero.
.TP
.BI \-o " number"
.I JBIG
separates an image into several horizontal stripes, resolution layers
and planes, were each plane contains one bit per pixel. One single
stripe in one plane and layer is encoded as a data unit called stripe
data entity (SDE) inside the BIE. There are 12 different possible
orders in which the SDEs can be stored inside the BIE and
.I number
selects which one shall be used. The order of the SDEs is only relevant
for applications that want to decode a
.I JBIG
file which has not yet completely arrived from e.g. a slow network connection.
For instance some applications prefer that the outermost of the three loops
(stripes, layers, planes) is over all layers so that all data of the lowest
resolution layer is transmitted first.
.br
The following values for
.I number
select these loop arrangements for writing the SDEs (outermost
loop first):

   0  	planes, layers, stripes
.br
   2  	layers, planes, stripes
.br
   3  	layers, stripes, planes
.br
   4  	stripes, planes, layers
.br
   5  	planes, stripes, layers
.br
   6  	stripes, layers, planes

All loops count starting with zero, however by adding 8 to the above
order code, the layer loop can be reversed so that it counts down to zero
and then higher resolution layers will be stored before lower layers.
Default order is 3 which writes at first all planes of the first
stripe and then completes layer 0 before continuing with the next
layer and so on. 
.TP
.BI \-p " number"
This option allows to activate or deactivate various optional algorithms
defined in the
.I JBIG
standard. Just add the numbers of the following options which you want to
activate in
order to get the
.I number
value:

   4 	deterministic prediction (DPON)
.br
   8 	layer 0 typical prediction (TPBON)
.br
  16 	diff. layer typ. pred. (TPDON)
.br
  64 	layer 0 two-line template (LRLTWO)

Except for special applications (like communication with
.I JBIG
subset implementations) and for debugging purposes you will normally
not want to change anything here. The default is 28, which provides
the best compression result.
.TP
.BI \-c
The adaptive template pixel movement is determined as suggested in
annex C of the standard. By default the template change takes place
directly in the next line which is most effective. However a few conformance
test examples in the standard require the adaptive template change to be
delayed until the first line of the next stripe. This option selects
this special behavior, which is normally not required except in order
to pass some conformance test suite.
.TP
.BI \-v
After the BIE has been created, a few technical details of the created
file will be listed (verbose mode).
.SH STANDARDS
This program implements the
.I JBIG
image coding algorithm as specified in ISO/IEC 11544:1993 and
ITU-T T.82(1993).
.SH AUTHOR
The
.I pbmtojbig
program is part of the 
.I JBIG-KIT
package, which has beed developed by Markus Kuhn.
The most recent version of this
portable
.I JBIG
library and tools set is freely available on the Internet
from anonymous ftp server ftp.uni-erlangen.de in directory pub/doc/ISO/JBIG/.
Bug reports should be sent to <mskuhn@cip.informatik.uni-erlangen.de>. 
.SH SEE ALSO
pbm(5), jbigtopbm(1)


CHANGES

JBIG-KIT revision history
-------------------------


Changes in version 2.1 (2014-04-08)

This is a security-critical bug-fix release that remains API and ABI
backwards compatible to version 2.0. Users who process BIE data from
untrusted sources should upgrade.

  - fixed a buffer-overflow vulnerability in the jbig.c decoder,
    reported by Florian Weimer (Red Hat): CVE-2013-6369

  - fixed ability of corrupted input data to force jbig85.c decoder
    into an end-less loop

  - fixed a bug in the processing of private deterministic-prediction
    tables (DPPRIV=1) in jbig.c decoder

  - fixed integer-type mismatches in printf arguments on 64-bit systems

  - fuzz-testing script added


Changes in version 2.0 (2008-08-30)

Main new features of release 2.0:

  - This release adds a separate lightweight "T.85" version of the
    encoder and decoder library (jbig85.c, jbig85.h). This new T.85
    library is optimized for embedded low-memory applications, such as
    printers and fax machines. It implements only the subset of the
    JBIG standard that is required by the ITU-T T.85 fax profile,
    namely only a single bit plane (P = 1) and no differential layers
    (DL = D = HITOLO = SEQ = ILEAVE = SMID = TPDON = DPON = DPON =
    DPLAST = 0) and some other restrictions (MY = 0, MX < 128).

    The T.85 library requires only three pixel rows of the
    uncompressed image to reside in memory at any time. This 3-line
    buffer has to be allocated by the calling application, therefore
    the T.85 library will not call any heap management functions such
    as malloc() or free() itself. It can fully deal with endless-paper
    fax, where the height of the encoded image is not known initially.

  - There are also two new example applications pbmtojbg85 and
    jbgtopbm85 included that provide command-line access to the new
    T.85 library.

API changes in 2.0 to the original full jbig.c library (compared to 1.6):

  - jbg_dec_getwidth(), jbg_dec_getheight(), jbg_dec_getsize(), and
    jbg_dec_getsize_merged() return now an unsigned long value
    (was: long), and return 0 (was: -1) if no image is available yet

  - jbg_dec_in() return values changed; the least significant bits of
    the error codes now provide more information about what caused the
    error, beyond what the more general error text messages returned by
    jbg_strerror() reveal

  - non-English translations of error messages are now provided
    in a separate PO file (e.g., for use with GNU gettext),
    jbg_strerror() no longer has a language attribute and the
    JBG_EN, JBG_DE_8859_1, JBG_DE_UTF_8 constants no longer exist

Other changes in 2.0 to the original full jbig.c library:

  - the arithmetic codec routines are now located in a separate source
    file (they are now used by both the full library jbig.c and the new
    T.85 library jbig85.c, and might also be of use to other applications)

  - arithmetic decoder behaviour slightly changed to enable the
    lookahead needed to handle T.85-style NEWLEN after the final
    stripe

  - added encoder support for adding comment marker segment

  - added encoder option for using SDRST marker instead of SDNORM
    (only useful to make encoder able to generate a wider range
    of test data)

  - macro JBG_LICENCE can be changed from the default value "GPL" to a
    licence agreement reference code to if the library is used under a
    commercial licence, to clarify under which exact licence the
    library is used in a particular application

Changes in 2.0 to the command-line wrappers for jbig.c:

  - added pbmtojbg options -C and -r, which provide command-line
    access to the previous two extensions

  - slightly improved diagnostic output of jbgtopbm option -d

  - pbmtojbg has a new option -f as a shortcut for all the options
    needed to make the output comply to the basic ITU-T T.85 fax
    profile

  - jbgtopbm now outputs image dimensions in header padded to 10
    digits fixed width, for consistency with jbgtopbm85
    to allow for later in-place update of height due to NEWLEN


Changes in version 1.6 (2004-06-11)

  - various small changes to reduce the risk of 32-bit unsigned
    integer overflows when dealing with extremely large images

  - robuster treatment of L0 = 0xffffffff

  - minor API modification in jbg_enc_options(): parameter l0 changed
    from type long to unsigned long; previous value now remains
    unchanged when l0 == 0 (was: l0 < 0)

  - lots of type casts added such that the C source code is now
    also compilable as C++


Changes in version 1.5 (2003-06-11)

  - fixed two minor memory leaks (special thanks to Davide Pizzolato
    <ing.davide.pizzolato@libero.it> for locating one of these)

  - jbgtopbm does not attempt any more to parse multiple concatenated
    BIEs (options -m must be used now to enable this feature explicitly),
    in order to handle BIEs with data after the last expected SDE gracefully

  - various extensions to improve support of JBIG fax applications
    (ITU-T T.85 profile):

      o support for maximum adaptive template pixel offset increased
        to MX=127 in both encoder and decoder

      o encoder now has a hook for producing BIEs with a NEWLEN marker
        segment and VLENGTH=1, in order to assist in testing decoders
        for T.85 conformance (see also new pbmtojbg option -Y)

      o a new function jbg_newlen() can be used to scan an
        entire in-memory BIE for NEWLEN marker segments and update the
        YD value in the BIE header, which can be applied to BIE data
        before passing it to the decoder for T.85 compliance
        (Background: the incremental design of the JBIG-KIT decoder
        makes it infeasible to look ahead for NEWLEN marker segments
        that occur after the SDE with the last image line, therefore
        this second pass is necessary to handle the old-style fax
        applications permitted by T.85 in which the low-RAM encoder

pbmtools/pbmtojbig.c
/*
 *  pbmtojbig - Portable Bitmap to JBIG converter
 *
 *  Markus Kuhn -- mskuhn@cip.informatik.uni-erlangen.de
 *
 *  $Id: pbmtojbg.c,v 1.2 1995-06-08 16:44:30 mskuhn Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include "jbig.h"

const char usage_msg[] = "PBMtoJBIG converter 0.6 -- "
"creates bi-level image entity (BIE) as output file\n\n"
"usage: %s [<options>] [<input-file> | -  [<output-file>]]\n\n"
"options:\n\n"
"  -q\t\tsequential coding, no differential layers (like -d 0)\n"
"  -x number\tmaximum width of lowest resolution layer (default 640)\n"
"  -y number\tmaximum height of lowest resolution layer (default 480)\n"
"  -l number\tlowest layer written to output file (default 0)\n"
"  -h number\thighest layer written to output file (default max)\n"
"  -d number\ttotal number of differential layers (overrides -x and -y)\n"
"  -s number\theight of a stripe in layer 0\n"
"  -m number\tmaximum adaptive template pixel horizontal offset (default 8)\n"
"  -o number\torder byte value: add 1=SMID, 2=ILEAVE, 4=SEQ, 8=HITOLO\n"
"\t\t(default 3 = ILEAVE+SMID)\n"
"  -p number\toptions byte value: add DPON=4, TPBON=8, TPDON=16, LRLTWO=64\n"
"\t\t(default 28 = DPON+TPBON+TPDON)\n"
"  -c\t\tdelay adaptive template changes to first line of next stripe\n"
"\t\t(only provided for a conformance test)\n"
"  -v\t\tverbose output\n\n";

char *progname;                  /* global pointer to argv[0] */
unsigned long total_length = 0;  /* used for determining output file length */


/* 
 * Read an ASCII integer number from file f and skip any PBM
 * comments which are encountered.
 */
unsigned long getint(FILE *f)
{
  int c;
  unsigned long i;

  while ((c = getc(f)) != EOF && !isdigit(c))
    if (c == '#')
      while ((c = getc(f)) != EOF && !(c == 13 || c == 10)) ;
  if (c != EOF) {
    ungetc(c, f);
    fscanf(f, "%lu", &i);
  }

  return i;
}


/*
 * Callback procedure which is used by JBIG encoder to deliver the
 * encoded data. It simply sends the bytes to the output file.
 */
void data_out(unsigned char *start, size_t len, void *file)
{
  fwrite(start, len, 1, (FILE *) file);
  total_length += len;
  return;
}


/*
 * Print usage message and abort
 */
void usage(void)
{
  fprintf(stderr, usage_msg, progname);
  exit(1);
}


int main (int argc, char **argv)
{
  FILE *fin = stdin, *fout = stdout;
  char *fnin = "<stdin>", *fnout = "<stdout>";
  int c, i, j;
  int all_args = 0, files = 0;
  unsigned long width, height;
  size_t bitmap_size;
  char type;
  unsigned char *bitmap, *p;
  struct jbg_enc_state s;
  int verbose = 0, delay_at = 0;
  long mwidth = 640, mheight = 480;
  int dl = -1, dh = -1, d = -1, l0 = -1, mx = -1;
  int options = JBG_TPDON | JBG_TPBON | JBG_DPON;
  int order = JBG_ILEAVE | JBG_SMID;


  /* parse command line arguments */
  progname = argv[0];
  for (i = 1; i < argc; i++) {
    if (!all_args && argv[i][0] == '-')
      if (argv[i][1] == '\0' && files == 0)
	++files;
      else
	for (j = 1; j < 10000 && argv[i][j]; j++)
	  switch(tolower(argv[i][j])) {
	  case '-' :
	    all_args = 1;
	    break;
	  case 'v':
	    verbose = 1;
	    break;
	  case 'c':
	    delay_at = 1;
	    break;
	  case 'x':
	    if (++i >= argc) usage();
	    j = 10000;
	    mwidth = atol(argv[i]);
	    break;
	  case 'y':
	    if (++i >= argc) usage();
	    j = 10000;
	    mheight = atol(argv[i]);
	    break;
	  case 'o':
	    if (++i >= argc) usage();
	    j = 10000;
	    order = atoi(argv[i]);
	    break;
	  case 'p':
	    if (++i >= argc) usage();
	    j = 10000;
	    options = atoi(argv[i]);
	    break;
	  case 'l':
	    if (++i >= argc) usage();
	    j = 10000;
	    dl = atoi(argv[i]);
	    break;
	  case 'h':
	    if (++i >= argc) usage();
	    j = 10000;
	    dh = atoi(argv[i]);
	    break;
	  case 'q':
	    d = 0;
	    break;
	  case 'd':
	    if (++i >= argc) usage();
	    j = 10000;
	    d = atoi(argv[i]);
	    break;
	  case 's':
	    if (++i >= argc) usage();
	    j = 10000;
	    l0 = atoi(argv[i]);
	    break;
	  case 'm':
	    if (++i >= argc) usage();
	    j = 10000;
	    mx = atoi(argv[i]);
	    break;
	  default:
	    usage();
	  }
    else
      switch (files++) {
      case 0:
	if (argv[i][0] != '-' || argv[i][1] != '\0') {
	  fnin = argv[i];
	  fin = fopen(fnin, "rb");
	  if (!fin) {
	    fprintf(stderr, "Can't open input file '%s", fnin);
	    perror("'");
	    exit(1);
	  }
	}
	break;
      case 1:
	fnout = argv[i];
	fout = fopen(fnout, "wb");
	if (!fout) {
	  fprintf(stderr, "Can't open input file '%s", fnout);
	  perror("'");
	  exit(1);
	}
	break;
      default:
	usage();
      }
  }
  
  /* read PBM header */
  while ((c = getc(fin)) != EOF && (isspace(c) || c == '#'))
    if (c == '#')
      while ((c = getc(fin)) != EOF && !(c == 13 || c == 10)) ;
  if (c != 'P') {
    fprintf(stderr, "Input file '%s' does not look like a PBM file!\n", fnin);
    exit(1);
  }
  type = getc(fin);
  width = getint(fin);
  height = getint(fin);
  getc(fin);    /* skip line feed */

  /* read PBM image data */
  bitmap_size = ((((width - 1) | 7) + 1) >> 3) * (size_t) height;
  bitmap = malloc(sizeof(unsigned char) * bitmap_size);
  if (!bitmap) {
    fprintf(stderr, "Sorry, not enough memory available!\n");
    exit(1);
  }
  p = bitmap;
  switch (type) {
  case '1':
    /* PBM text format */
    for (i = 0; i < height; i++)
      for (j = 0; j < width | 7; j++) {
	*p <<= 1;
	if (j < width)
	  *p |= getint(fin) & 1;
	if ((j & 7) == 7)
	  ++p;
      }
    break;
  case '4':
    /* PBM raw binary format */
    fread(bitmap, bitmap_size, 1, fin);
    break;
  default:
    fprintf(stderr, "Unsupported PBM type P%c!\n", type);
    exit(1);
  }
  if (ferror(fin)) {
    fprintf(stderr, "Problem while reading input file '%s", fnin);
    perror("'");
    exit(1);
  }
  if (feof(fin)) {
    fprintf(stderr, "Unexpected end of input file '%s'!\n", fnin);
    exit(1);
  }

  /* Apply JBIG algorithm and write BIE to output file */

  /* initialize parameter struct for JBIG encoder*/
  jbg_enc_init(&s, width, height, 1, &bitmap, data_out, fout);

  /* Select number of resolution layers either directly or based
   * on a given maximum size for the lowest resolution layer */
   if (d >= 0)
    jbg_enc_layers(&s, d);
  else
    jbg_enc_lrlmax(&s, mwidth, mheight);

  /* Specify a few other options (each is ignored if negative) */
  if (delay_at)
    options |= JBG_DELAY_AT;
  jbg_enc_lrange(&s, dl, dh);
  jbg_enc_options(&s, order, options, l0, mx, -1);

  /* now encode everything and send it to data_out() */
  jbg_enc_out(&s);

  /* give encoder a chance to free its temporary data structures */
  jbg_enc_free(&s);

  /* check for file errors and close fout */
  if (ferror(fout) || fclose(fout)) {
    fprintf(stderr, "Problem while writing output file '%s", fnout);
    perror("'");
    exit(1);
  }

  /* In case the user wants to know all the gory details ... */
  if (verbose) {
    fprintf(stderr, "Information about the created JBIG bi-level image entity "
	    "(BIE):\n\n");
    fprintf(stderr, "              input image size: %ld x %ld pixel\n",
	    s.xd, s.yd);
    fprintf(stderr, "                    bit planes: %d\n", s.planes);
    fprintf(stderr, "                       stripes: %ld\n", s.stripes);
    fprintf(stderr, "   lines per stripe in layer 0: %ld\n", s.l0);
    fprintf(stderr, "  total number of diff. layers: %d\n", s.d);
    fprintf(stderr, "           lowest layer in BIE: %d\n", s.dl);
    fprintf(stderr, "          highest layer in BIE: %d\n", s.dh);
    fprintf(stderr, "             lowest layer size: %lu x %lu pixel\n",
	    ceil_half(s.xd, s.d - s.dl), ceil_half(s.yd, s.d - s.dl));
    fprintf(stderr, "            highest layer size: %lu x %lu pixel\n",
	    ceil_half(s.xd, s.d - s.dh), ceil_half(s.yd, s.d - s.dh));
    fprintf(stderr, "                   option bits:%s%s%s%s%s%s%s\n",
	    s.options & JBG_LRLTWO  ? " LRLTWO" : "",
	    s.options & JBG_VLENGTH ? " VLENGTH" : "",
	    s.options & JBG_TPDON   ? " TPDON" : "",
	    s.options & JBG_TPBON   ? " TPBON" : "",
	    s.options & JBG_DPON    ? " DPON" : "",
	    s.options & JBG_DPPRIV  ? " DPPRIV" : "",
	    s.options & JBG_DPLAST  ? " DPLAST" : "");
    fprintf(stderr, "                    order bits:%s%s%s%s\n",
	    s.order & JBG_HITOLO ? " HITOLO" : "",
	    s.order & JBG_SEQ    ? " SEQ" : "",
	    s.order & JBG_ILEAVE ? " ILEAVE" : "",
	    s.order & JBG_SMID   ? " SMID" : "");
    fprintf(stderr, "           AT maximum x-offset: %d\n"
	    "           AT maximum y-offset: %d\n", s.mx, s.my);
    fprintf(stderr, "         length of output file: %lu byte\n\n",
	    total_length);
  }
  return 0;
}
